smart_card_connector_app/src/chrome-api-jstocxxtest.js

```diff
@@ -640,6 +659,139 @@ goog.exportSymbol('testChromeApiProviderToCpp', {
             'onDisconnectRequested', /*requestId=*/ 113, BAD_HANDLE,
             chrome.smartCardProviderPrivate.Disposition.LEAVE_CARD)
         .$waitAndVerify();
+  },
+
+  'testConnectedReader': {
+    'setUp': async function() {
+      launchPcscServer(/*initialDevices=*/[{
+        'id': 123,
+        'type': SimulationConstants.GEMALTO_DEVICE_TYPE,
+        'cardType': SimulationConstants.COSMO_CARD_TYPE
+      }]);
+      createChromeApiProvider();
+      const sCardContext = await establishContext(/*requestId=*/ 1111);
+      readerHandle = await connect(
+          /*requestId=*/ 1112, sCardContext,
+          SimulationConstants.GEMALTO_PC_TWIN_READER_PCSC_NAME0,
+          chrome.smartCardProviderPrivate.ShareMode.SHARED,
+          /*preferredProtocols=*/ {'t1': true},
+          /*expectedProtocol=*/ chrome.smartCardProviderPrivate.Protocol.T1);
+    },
+
+    'tearDown': async function() {
+      expectReportPlainResult(/*requestId=*/ 2222, 'SUCCESS');
+      await mockChromeApi
+          .dispatchEvent(
+              'onDisconnectRequested', /*requestId=*/ 2222, readerHandle,
+              chrome.smartCardProviderPrivate.Disposition.LEAVE_CARD)
+          .$waitAndVerify();
+    },
+
+    // Test that Status returns information about the card.
+    'testStatus': async function() {
+      expectReportStatusResult(
+          /*requestId=*/ 111,
+          SimulationConstants.GEMALTO_PC_TWIN_READER_PCSC_NAME0,
+          chrome.smartCardProviderPrivate.ConnectionState.NEGOTIABLE,
+          chrome.smartCardProviderPrivate.Protocol.T1,
+          SimulationConstants.COSMO_ID_70_ATR, 'SUCCESS');
+      await mockChromeApi
+          .dispatchEvent('onStatusRequested', /*requestId=*/ 111, readerHandle)
+          .$waitAndVerify();
+    },
+
+    // Test that Status returns READER_UNAVAILABLE after reader disappears.
+    'testStatus_readerUnavailable': async function() {
+      setSimulatedDevices(/*devices=*/[]);
+
+      // Status should eventually return a READER_UNAVAILABLE error.
+      for (let requestId = 111; ; requestId += 1) {
```

Comment:
Having such a busy loop is unfortunate. It's unclear how many times it spins in a test run (if any), whether it's crucial for test coverage that it spins at least once. Also if it happens to work sufficiently long, `requestId` will start coinciding with those used in the setUp/tearDown.

Can we have a more deterministic test here? E.g., will we lose on test coverage if we launch the PC/SC server without any devices, do a single `onStatusRequested` and verify it returns `READER_UNAVAILABLE`?
Alternatively, if you really want to test the transition, you can use `SCardGetStatusChange()` or analogs to wait for it explicitly.

Reply:
> It's unclear how many times it spins in a test run

around 50
> E.g., will we lose on test coverage if we launch the PC/SC server without any devices, do a single onStatusRequested and verify it returns READER_UNAVAILABLE

it won't work like this, since you first need to make a `Connect` call.

>Alternatively, if you really want to test the transition, you can use SCardGetStatusChange() or analogs to wait for it explicitly.

I'll check if smth like this works. Initially I just made a similar test to the one in `application_unittest.cc` and there it's claimed that `(we have to do a polling loop since there's no simple way to observe this asynchronous event).`

Reply:
OK indeed I forgot about that C++ test that I wrote :D Alright, in that case I'd say - it's fine to keep the polling in the JS as well, but please fix the request id overlap.

Reply:
fixed the request id overlap